Fixed issues 4 locking full posts pages content with new refactored unlock protocol flow

SolomonFoskaay opened the pull request:
# What was done?
## 3 Important Things:
**1.** Refactored the current Unlock Protocol Flow used by the plugin (removed Class And Singleton design which complicated the process of extending the plugin with new features). This makes it far more easier for contributing developers to add new features to the plugin moving foward.

**2.** Fixed issue #4 which means the Unlock Protocol plugin can now lock full post/page content with the same Unlock protocol flow used previously for locking Gutenberg block content with Unlock-based NFTs.

**3.** Made the no1 changes without breaking the existing codebase which still allows the Gutenberg Block lock to function as before while the full post/page content lock work side by side.

SolomonFoskaay committed:
Added version2 and full-post-page content lock options to post/page editor sidebar

SolomonFoskaay committed:
Added full-post-page content lock options to post/page editor sidebar

SolomonFoskaay committed:
fixing unlock protocol flow for full post lock

SolomonFoskaay committed:
fixed unlock protocol flow in full post/page content lock

SolomonFoskaay committed:
fixed unlock protocol flow in full post/page content lock

SolomonFoskaay committed:
Added full-post-page content lock options to post/page editor sidebar + new refactored unlock protocol flow

julien51 commented:
> (removed Class And Singleton design which complicated the process of extending the plugin with new features)

As noted several times @SolomonFoskaay I am confident this design was the right one. It was done by a team that was recommended by Automattic and which I am confident are WP experts. 
The issue and bounty was about adding the ability to lock posts, not about refactoring the plugin.

SolomonFoskaay committed:
Added full post/page content lock with Unlock Protocol

SolomonFoskaay commented:
@julien51  kindly check it out now. Fully integrated into the existing code base as requested.
Thanks

SolomonFoskaay committed:
Hide save lock button and don't save locks when no lock added, empty network, empty lock or invalid lock address added

julien51 commented on unlock-wordpress-plugin/inc/classes/class-unlock.php:
```diff
@@ -47,6 +47,8 @@ public static function get_checkout_url_base() {
 		return $checkout_url_base;
 	}
 
+
+
```
```suggestion
```

julien51 committed:
Update unlock-wordpress-plugin/inc/classes/class-unlock.php

julien51 commented on unlock-wordpress-plugin/unlock-protocol.php:
```diff
@@ -36,3 +36,4 @@ function unlock_protocol_plugin_loader() {
 }
 
 unlock_protocol_plugin_loader();
+
```
```suggestion
```

julien51 committed:
Update unlock-wordpress-plugin/unlock-protocol.php

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-editfullpp.js:
```diff
@@ -0,0 +1,323 @@
+document.addEventListener('DOMContentLoaded', function () {
```
Is this standard, I am surprised you would need to do that...

julien51 replied:
We should use Gutember blocks for all this. Please check how the plugin does it for the "blocks".

SolomonFoskaay replied:
fixed by using reacjs in the "src" folder

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-editfullpp.php:
```diff
@@ -0,0 +1,119 @@
+        echo '<p><div id="lock-list"></div></p>';
+        echo '<p><div id="feedback"></div></p>';
+        echo '<p><button id="save-lock" type="button" class="button" style="background-color: green; color: white;">Save Lock</button></p>';
+        echo '</div>';
```
This does not feel right here. You should not create HTML like this. Gutemberg uses react to create the components. Let's use this here. Check `src/assets/js` for example of how the plugin does it.

SolomonFoskaay replied:
Fixed: I have deleted this file, and recreated the functions by using reactjs in the "src" folder

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,196 @@
+		}, array_values($networks));
+
+		// Check if the attributes are valid
+		if (!$attributes || !isset($attributes['locks']) || !isset($attributes['ethereumNetworks'])) {
+			return $content;
+		}
```
I think we should not render the content here but only do it from `render_fullpp_content`

SolomonFoskaay replied:
It is essential to render content if lock(s) is not added at all in the post/page because this is different from the existing block implementation, this deals with the "the_content" hook which is the full post/page content wp hook. 
If lock(s) are not added, it won't be able to trigger the unlock Protocol NFT flow and at the same time, it will hide the content from displaying at all in frontend (even for admin).

That's why it is better to have it so that the content can be shown in case unlock protocol NFT flow is not triggered because no lock(s) is added.

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,196 @@
+		}
+
+
+		// Step 4: Reconstruct a new attributes format to match the old structure
```
I am not sure I understand what "the old structure" is?

SolomonFoskaay replied:
Please don't worry about this, the function has been removed entirely in the new PR pushed below.

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,196 @@
+	 *
+	 * @return mixed|void
+	 */
+	private function get_checkout_url( $attributes ) {
```
This does not reallyt get the checkout URL but render the checkout content, so let's rename to `render_checkout_content`

julien51 replied:
We should also re-use what the plugin already does for blocks.

SolomonFoskaay replied:
Yeah, all the render_block and get_checkout_url functions codes and the naming convention are used as it was used 100% in the block, I did not add or remove a code from it.

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,196 @@
+			! is_user_logged_in() ||
+			( is_user_logged_in() && ! up_get_user_ethereum_address() )
+		) {
+			$login_button_text       = up_get_general_settings( 'login_button_text', __( 'Login with Unlock', 'unlock-protocol' ) );
```
Let's move this to its own function `render_login_content`... or even better, let's just reuse what the rest of the plugin already does?

SolomonFoskaay replied:
fixed. used as existing code.

julien51 reviewed (commented):
It is better but you are still not using the same approach as the rest of the plugin for all the JS. Pleaase use Gutemberg blocks as this is the way to go in WordPress!

SolomonFoskaay committed:
Added full-post-page content lock options to post/page editor sidebar

SolomonFoskaay commented:
> It is better but you are still not using the same approach as the rest of the plugin for all the JS. Pleaase use Gutemberg blocks as this is the way to go in WordPress!

I have followed the existing Unlock protocol Wordpress codebase using the Reactjs + Class + Singleton design pattern in this newly submitted PR.

Kindly check it out.
Thanks!

julien51 reviewed (commented):
This is starting to look good! Let me do a thorough review tomorrow as I will need to checkout the code locally!

SolomonFoskaay commented:
> This is starting to look good! Let me do a thorough review tomorrow as I will need to checkout the code locally!

Okay, awaiting this. 
Thanks!

julien51 commented:
Thanks a lot @SolomonFoskaay !

Here are a few notes from the UI.

First there is still a UI issue on the login page:

<img width="499" alt="Screenshot 2023-04-07 at 9 28 06 AM" src="https://user-images.githubusercontent.com/17735/230616980-c3e8b108-b513-4392-8b75-25a94a9295b7.png">

The config block has alignment issues:

<img width="310" alt="Screenshot 2023-04-07 at 9 33 26 AM" src="https://user-images.githubusercontent.com/17735/230617651-caeeebfd-bee1-45f1-a39b-3a470cb4fe3e.png">

Please move the link to Unlock's docs _outside_ the `Locks` section (can be added at the top with the description IMO).
The `Save` lock button should be added besides the `Remove` button.
<img width="292" alt="Screenshot 2023-04-07 at 9 35 16 AM" src="https://user-images.githubusercontent.com/17735/230617892-cf20908e-42f9-4c23-b505-28ad26d3d191.png">

And similarly the confirmation should be _inside_ the `Locks` box.
<img width="306" alt="Screenshot 2023-04-07 at 9 36 20 AM" src="https://user-images.githubusercontent.com/17735/230618028-80266f3c-dbe7-46a9-b85c-c157ed267d66.png">

clicking on `Remove` when a lock has not been saved does not work (it should just remove the extra form).

<img width="273" alt="Screenshot 2023-04-07 at 9 36 52 AM" src="https://user-images.githubusercontent.com/17735/230618111-55a108ac-1375-471d-bd90-47dea66c2564.png">

julien51 commented on package.json:
```diff
@@ -12,9 +12,11 @@
     "start": "wp-scripts start ./src/assets/js/main.js --output-path=./unlock-wordpress-plugin/assets/build/js",
     "start:blocks": "wp-scripts start ./src/assets/js/blocks.js --output-path=./unlock-wordpress-plugin/assets/build/js",
     "build:blocks": "wp-scripts build ./src/assets/js/blocks.js --output-path=./unlock-wordpress-plugin/assets/build/js",
+    "start:fullpostpage": "wp-scripts start ./src/assets/js/fullpostpage.js --output-path=./unlock-wordpress-plugin/assets/build/js",
```
Instead of introducing new tasks here I think it makes more sense to just re-use the `blocks` tasks and rename them into `public` (as opposed to the `admin` section)?

SolomonFoskaay replied:
have adjusted this to ensure the blocks' existing code is reused to handle both instead of introducing new one

julien51 commented on src/assets/js/fullpostpage.js:
```diff
@@ -0,0 +1,8 @@
+/**
```
Let's use `full-post-page.js` for the file name here for consistency with other files.

SolomonFoskaay replied:
okay. done.

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+import { __ } from "@wordpress/i18n";
```
Same here, let's use `edit-full-pust-page.js` (same with the folder)

SolomonFoskaay replied:
okay done.

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+import { useSelect } from '@wordpress/data';
+import "../../../scss/admin/editor.scss";
+
+const EditFullPP = ({ locks, ethereumNetworks, onUpdateLocks, onUpdateEthereumNetworks }) => {
```
```suggestion
const EditFullPostPage = ({ locks, ethereumNetworks, onUpdateLocks, onUpdateEthereumNetworks }) => {
```

SolomonFoskaay replied:
Updated already with latest PR push.

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+        fetchLockSettings(getpost.id);
+      })
+      .catch((err) => { });
+  }, [getpost.id, refresh]); // refresh the useEffect anytime either of this changes
```
The networks don't change much so I think it is ok to just fetch this once and not when`getpost.id` changes

SolomonFoskaay replied:
This is not just network, it required keeping the previous lock(s) setup be persistently displayed to the user when publish/update post/page button clicked or when the editor is fully reloaded. or else the user will not be able to interact with previously set lock(s).
The implementation is as needed to work with the post/page meta options panel sidebar.
Kindly consider this resolved too.

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+
+
+  // Check if the Ethereum address is valid
+  const isAddressValid = (address) => {
```
Maybe move this to a utils file? I suspect we already have a version of this existing in the plugin?

SolomonFoskaay replied:
This is a check that is needed when the reactjs is transpiled to js in the plugin folder so shouldn't be separated.
Kindly check the block implementation, it was used directly too.
So, this can be considered fixed.

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+
+    try {
+      const response = await apiFetch({
+        path: "/unlock-protocol/v1/save_unlockpfullpp_attributes",
```
Plese update the file name  for readability and consistency!

SolomonFoskaay replied:
name updated. fixed

julien51 commented on src/assets/js/fullpostpage/unlock-content/edit-fullpp.js:
```diff
@@ -0,0 +1,302 @@
+        method: "POST",
+        data: {
+          post_id: post_id,
+          unlockpfullpp_attributes: JSON.stringify(attributes),
```
Again, let's use more readable variable names here.

`unlockpfullpp_attributes` => `unlock_plugin_full_page_post_attributes`

SolomonFoskaay replied:
okay, done.

julien51 commented on unlock-wordpress-plugin/assets/build/css/fullpostpage.css:
```diff
@@ -0,0 +1 @@
+.unlock-header-icon{background-image:url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns=%27http://www.w3.org/2000/svg%27 fill=%27%23FF6771%27 height=%2720%27 width=%2720%27%3E%3Cpath fill-rule=%27evenodd%27 clip-rule=%27evenodd%27 d=%27M12.522 11.121a3.64 3.64 0 0 0 .032-.5l.001-1.724v-.432H7.577a3914.548 3914.548 0 0 0 0 2.225c.001 1.083.656 1.985 1.702 2.35 1.43.5 3.044-.454 3.243-1.919zm2.981- ...
```
Was this auto-generated? If so let's not add it...

julien51 replied:
anything in `build` should _not_ be added to `git`.

SolomonFoskaay replied:
This is confusing because the existing block and other implementations have their auto-generated files added to git. so wondering why this shouldn't be added for the new features I got added.

Anyways have included in the .gitignore file to henceforth ignore all files in the `build` folder and my latest PR does not include the full post/page auto-generated files.
So, consider this done too

julien51 commented on unlock-wordpress-plugin/inc/classes/class-fullpostpage.php:
```diff
@@ -0,0 +1,96 @@
+<?php
```
Please updare the file name!

SolomonFoskaay replied:
I did change the name as suggested above and it keeps breaking the plugin. So, I reverted it.
It is a reminder that when dealing with Singleton design patterns you don't have the freedom to adjust basic things like the file/variable names as you like. 

Did you notice I changed all the names as requested in the "src" because it's reactjs and is not directly connected or controlled by the Singleton traits in the plugin itself? 

So, I tried my best and was able to enforce some name changes as you desire but to enforce all the name changes, you will need to work on updating the Singleton "traits" to allow it and that will likely tamper with the existing codebase and features which have avoided base on your instructions previously.

So, I manage to adjust names that are adjustable without breaking the plugin and revert those that could not. Though those names are still okay to know what the feature is and not confusing.

So this fixd too

julien51 commented on unlock-wordpress-plugin/inc/classes/class-fullpostpage.php:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @since 3.0.0
+ */
+class Fullpostpage {
```
```suggestion
class FullPostPage {
```

SolomonFoskaay replied:
okay done. Though had to sacrifice not been able to change some other names to implement this due to the Singleton design pattern traits.

julien51 commented on unlock-wordpress-plugin/inc/classes/class-plugin.php:
```diff
@@ -35,6 +35,7 @@ protected function __construct() {
 		Blocks::get_instance();
 		Menu::get_instance();
 		API::get_instance();
+		Fullpostpage::get_instance();
```
```suggestion
		FullPostPage::get_instance();
```

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,172 @@
+		add_filter( 'the_content', array( $this, 'trigger_unlockprotocol_flow' ) );	
+    }
+	
+	/**
```
Formatting is strange here. Please use consistent spaces

SolomonFoskaay replied:
Not sure how that come to be may be old commit but the new one is well aligned. kindly recheck

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/class-unlock-box-fullpp.php:
```diff
@@ -0,0 +1,172 @@
+		if ( ! empty( $attributes ) ) {
+			// Call render_block() to trigger the Unlock Protocol flow.
+			return $this->render_block( $attributes, $content );
+		}else{
```
formatting

SolomonFoskaay replied:
Same here, well aligned. see uptodate PR please

julien51 commented on unlock-wordpress-plugin/inc/classes/fullpostpage/index.php:
```diff
@@ -0,0 +1,2 @@
+<?php
+// Silence is golden.
```
:D

SolomonFoskaay replied:
Lol... its a security measure to ensure the folder I added is not accessible directly from the browser. But seems you don't like it there. Have removed it completely considering is not part of the features been implemented in the issue..

SolomonFoskaay commented:
> First there is still a UI issue on the login page:
> 
> <img alt="Screenshot 2023-04-07 at 9 28 06 AM" width="499" src="https://user-images.githubusercontent.com/17735/230616980-c3e8b108-b513-4392-8b75-25a94a9295b7.png">

@julien51 have responded on this before that the issue is outside the scope of this issue. It defaults based on the existing design whenever the plugin is accessed in localwp. If you access the same dev plugin version in a production environment, it does not have that issue.

You can confirm this by trying the current "production" version of the Unlock plugin (not the dev version) in the localwwp. You will notice the same error. This means it's not specific to the dev version am working on but an existing issue in the codebase itself having conflict in localwp environment.

So, please I won't be working on it as part of this issue because it isn't related to it and is not caused by the new full post/page content-locking feature am introducing to the code base in this issue #4 .

**NOTE:** Aside from this, other comments you may are well noted and I will surely work on them all and push the update to this PR once done.

Thanks!

julien51 reviewed (commented):
Ok I added a bunch of comments.
Please take care of them and we'll do a final review.

Please make sure you use some linting, especially for the PHP files so we have consistent styling.

SolomonFoskaay commented:
> Ok I added a bunch of comments. Please take care of them and we'll do a final review.
> 
> Please make sure you use some linting, especially for the PHP files so we have consistent styling.

@julien51 Not sure I understand the part about using the linting. kindly help guide me further because I try to ensure I follow the existing codebase style as much as possible.

Thanks!

SolomonFoskaay committed:
Full post page all 5 UI issues fixed

SolomonFoskaay committed:
Fixed all 21 comments issues for the full post page lock feature

SolomonFoskaay commented:
> Thanks a lot @SolomonFoskaay !
> 
> Here are a few notes from the UI.
> 
> First there is still a UI issue on the login page:
> 
> <img alt="Screenshot 2023-04-07 at 9 28 06 AM" width="499" src="https://user-images.githubusercontent.com/17735/230616980-c3e8b108-b513-4392-8b75-25a94a9295b7.png">
> 
> The config block has alignment issues:
> 
> <img alt="Screenshot 2023-04-07 at 9 33 26 AM" width="310" src="https://user-images.githubusercontent.com/17735/230617651-caeeebfd-bee1-45f1-a39b-3a470cb4fe3e.png">
> 
> Please move the link to Unlock's docs _outside_ the `Locks` section (can be added at the top with the description IMO). The `Save` lock button should be added besides the `Remove` button. <img alt="Screenshot 2023-04-07 at 9 35 16 AM" width="292" src="https://user-images.githubusercontent.com/17735/230617892-cf20908e-42f9-4c23-b505-28ad26d3d191.png">
> 
> And similarly the confirmation should be _inside_ the `Locks` box. <img alt="Screenshot 2023-04-07 at 9 36 20 AM" width="306" src="https://user-images.githubusercontent.com/17735/230618028-80266f3c-dbe7-46a9-b85c-c157ed267d66.png">
> 
> clicking on `Remove` when a lock has not been saved does not work (it should just remove the extra form).
> 
> <img alt="Screenshot 2023-04-07 at 9 36 52 AM" width="273" src="https://user-images.githubusercontent.com/17735/230618111-55a108ac-1375-471d-bd90-47dea66c2564.png">

@julien51 All the above 5 issues mentioned in the UI are now fixed + the whole 16 code comments are fixed and marked resolved too (A total of about 21 fixes).

Meanwhile, there are about 2-3 folders/files I had to retain the name as it was before unchanged due to the following reasons:
I did change the name as suggested above and it keeps breaking the plugin. So, I reverted it.
It is a reminder that when dealing with Singleton design patterns you don't have the freedom to adjust basic things like the file/variable names as you like. 

Did you notice I changed all the names as requested in the "src" without leaving any unchanged because it's reactjs and is not directly connected or controlled by the Singleton traits in the plugin itself? 

So, I tried my best and was able to enforce some name changes as you desire but to enforce all the name changes, you will need to work on updating the Singleton "traits" to allow it and that will likely tamper with the existing codebase and features which have avoided doing myself base on your instructions previously.

So, I manage to adjust names that are adjustable without breaking the plugin and revert those that could not. Though even those unchanged names are still okay to know what the feature added is and not confusing in any way IMO.

Looking forward to having this as a final review possibly.

Thanks!

julien51 commented:
> @julien51 Not sure I understand the part about using the linting. kindly help guide me further because I try to ensure I follow the existing codebase style as much as possible.

When looking at your code I see indents all over the place for example. I think there are tools out there to make sure we have a "good looking" code that's easier to review and maintain.

julien51 reviewed (commented):
ok After a quick look atthe code I think this is all pretty good! I will do some extensive Q&A testing a bit later today but if everything works as expected we're done! 
I'll confirm shortly!
Thanks!

julien51 commented:
Hey so there are still issues:

The plugin is now in version 4, which is ok, but we still show this: `3.2.3`
<img width="651" alt="Screenshot 2023-04-13 at 11 09 22 AM" src="https://user-images.githubusercontent.com/17735/231803813-23887967-1148-438e-b5fa-a69f5b5b07c3.png">

I still have this warning:
<img width="475" alt="Screenshot 2023-04-13 at 11 09 47 AM" src="https://user-images.githubusercontent.com/17735/231803950-5a532494-75a1-435d-b2ef-0bb63a9a6ef3.png">

On the editor I am unable to see the option to lock a post anymore:

<img width="345" alt="Screenshot 2023-04-13 at 11 11 22 AM" src="https://user-images.githubusercontent.com/17735/231804412-270bf292-ec0b-4cb1-80a3-1948f5bfd7e1.png">

julien51 commented:
What I did locally is pulled your latest changes, then run `yarn release` (lots of warning... see below) and installed the plugin as a new plugin on the Wordpress install locally:


It would be good to clear out these warnings:

```
Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246
PHP Deprecated:  Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 245

Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 245
PHP Deprecated:  Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246

Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246
Success: POT file successfully generated!

> unlock-protocol-plugin@4.0.0 zip
> node zip

Copying files...
cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/inc
cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/templates
cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/unlock-protocol.php
sed: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/dist/unlock-wordpress-plugin/unlock-protocol.php
Making zip...
Done.
✨  Done in 7.69s.
```

SolomonFoskaay commented:
> I still have this warning:
> <img alt="Screenshot 2023-04-13 at 11 09 47 AM" width="475" src="https://user-images.githubusercontent.com/17735/231803950-5a532494-75a1-435d-b2ef-0bb63a9a6ef3.png">

@julien51 **HINT:** Use the `tab key` to navigate the "Wordpress login page" on localwp. As I said, the issue on the login page is an existing issue with the plugin codebase and if you test the non-dev version of the plugin, you will see the same error. Once you use that, you will be able to login into localwp successfully. 

Meanwhile, `you keep raising this issue which I have responded to that severally as well that the error is from the existing plugin code base and not codes I introduced`. Since it seems you trusted the previous developers not to make those errors and seem not too convinced yet about me. Kindly try the existing non-developer version plugin currently used by the community, install it in localwp and try it, you will see the same error (and I did not work on that version of the plugin at all) to confirm this is not issue caused by me working on the plugin.

SolomonFoskaay commented:
> The plugin is now in version 4, which is ok, but we still show this: `3.2.3`
> <img alt="Screenshot 2023-04-13 at 11 09 22 AM" width="651" src="https://user-images.githubusercontent.com/17735/231803813-23887967-1148-438e-b5fa-a69f5b5b07c3.png">

@julien51 I am confused about this, I only worked on the developer version of the plugin and did not update anything about the community version. Kindly explain how this relates to the issue #4 I fixed, please.

Meanwhile, `based on the version number mismatch, it's because you have not changed the description of the plugin to the 4 you intend to release inside the main plugin file`. Once you change it there, it will update.

My question is are all these part of the issue #4 I was assigned? They are not, and in fairness, bringing them in to delay approving the fixes will not help have this issue approved. 

PLEASE, I AM pleading, let other issues that are not part of this issue not be introduced to cause a delay in its approval again this time. PLEASE!

SolomonFoskaay commented:
> On the editor I am unable to see the option to lock a post anymore:
> 
> <img alt="Screenshot 2023-04-13 at 11 11 22 AM" width="345" src="https://user-images.githubusercontent.com/17735/231804412-270bf292-ec0b-4cb1-80a3-1948f5bfd7e1.png">

@julien51 that is not displaying because you told me to remove anything in the `build` folder and not to be pushed to git. `And I asked you why? When the existing codebase does not exclude it. So, I excluded them in the new PR I sent to abide by your instruction even though I knew it could cause issues like this.`
See your instruction I followed about it from the previous 20+ comments I fixed below:
![Selection_085](https://user-images.githubusercontent.com/83863629/231820333-b36b9649-6a47-4d62-8b14-52b7e66e00ec.png)

## How to fix: 
Since the build folder js and CSS files are not included in the PR anymore, you will always have to manually add them whenever you (or any contributing developer) want to work with the developer version for the Unlock plugin.

NOTE: And you also have to ensure it's rebuilt on your local dev environment before you attempt to build a production version with `yarn release` or else the same breaking issues will happen during the release build stage).

**STEP 1:** Open the `unlock-wordpress-plugin` main folder in your code editor (Vscode etc.)

**STEP 2:** Open `terminal` and `yarn`. When done. Build the excluded `build` folder and this uses webpack to transpile the  `reactjs` and `scss` files in the `src` folder to pure `js` and `css` in the `build` folder of the actual plugin at `unlock-wordpress-plugin/unlock-wordpress-plugin/assets/build`.

Once you follow the above instructions, all the build folder files essential to run the plugin will be added, then you can try using it and you will see the missing `lock` options back and working fine.

`To make things easier for you because I know you don't have much time, I have decided to remove the .gitignore code and reinclude the build folder back as it was before you asked me to exclude it from git.`

So, kindly retry the new PR, I will push after this response and you won't have the issues and won't need to follow the above steps to rebuild the excluded build folder by yourself again.

NOTE: I always try my best to ensure the best for this work but seems you always want things done a particular way and to avoid a further argument, I decided to always follow your path even when I knew it could cause issues down the line just like this issue. `Please, my humble advice is, It will be better when working with devs to at least give some space for the dev working directly on the issue to do some thing a bit differently from how you expect as long as it does not jopadice the security and user experience`.

Thanks!

SolomonFoskaay committed:
Re-added previously excluded build folder causing breaking issues

SolomonFoskaay commented:
> What I did locally is pulled your latest changes, then run `yarn release` (lots of warning... see below) and installed the plugin as a new plugin on the Wordpress install locally:
> 
> It would be good to clear out these warnings:
> 
> ```
> Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246
> PHP Deprecated:  Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 245
> 
> Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 245
> PHP Deprecated:  Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246
> 
> Deprecated: Use of "static" in callables is deprecated in phar:///opt/homebrew/Cellar/wp-cli/2.7.1/bin/wp/vendor/wp-cli/i18n-command/src/IterableCodeExtractor.php on line 246
> Success: POT file successfully generated!
> 
> > unlock-protocol-plugin@4.0.0 zip
> > node zip
> 
> Copying files...
> cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/inc
> cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/templates
> cp: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/unlock-wordpress-plugin/unlock-protocol.php
> sed: no such file or directory: /Users/julien/repos/unlock-wordpress-plugin/dist/unlock-wordpress-plugin/unlock-protocol.php
> Making zip...
> Done.
> ✨  Done in 7.69s.
> ```

@julien kindly test the new PR in localwp developer first, once you confirm everything about the 20+ comments fixed and working.

We can now work together on issues outside of this issue #4  . All the above issues are outside what I worked on in this issue 4 and please let's stop bringing them in or else things will mix up and won't progress on this issue..

Thanks!

julien51 committed:
minor UI tweak

julien51 committed:
Merge branch 'UPWPlugin08a-Fixed-Issues-4-Locking-Full-Posts-Pages-with-new-UnlockP-flow' of github.com:SolomonFoskaay/unlock-wordpress-plugin into UPWPlugin08a-Fixed-Issues-4-Locking-Full-Posts-Pages-with-new-UnlockP-flow

julien51 commented:
@SolomonFoskaay Please remove `build` files again. As I explained to you this is a horrible software development practice.

SolomonFoskaay committed:
Removed all build folder files from git commits

SolomonFoskaay commented:
> @SolomonFoskaay Please remove `build` files again. As I explained to you this is a horrible software development practice.

@julien51 Yeah I reason with what you said, it is okay to exclude the `build` folder files. So, I did the following to the new PR I just pushed above:

1. Git pull all changes to the master branch and your UI adjustment to this PR. I noticed new changes have been merged to the plugin master branch and pull them in before I make the requested adjustment below to ensure this PR is up-to-date with master.

2. Deleted all `build` folder files (I deleted `build` files including those existing there already before I worked on this plugin at all and the ones included when I added the full post/page content lock features)

3. Added instruction for git to ignore committing `build` folder files moving forward in the `.gitignore` file

4. Before I sent the new PR

So, kindly git pull the update to this PR now and check it out.

Thanks!

julien51 commented on src/assets/js/full-post-page/unlock-content/edit-full-post-page.js:
```diff
@@ -0,0 +1,282 @@
+  }, []);
+
+  //check if current post is already 
+  const { isSavingPost, didPostSaveRequestSucceed } = useSelect((select) => {
```
I don't understand this and I am not sure this is needed.

julien51 commented on src/assets/js/full-post-page/unlock-content/edit-full-post-page.js:
```diff
@@ -0,0 +1,282 @@
+const EditFullPostPage = ({ locks, ethereumNetworks, onUpdateLocks, onUpdateEthereumNetworks }) => {
+
+  const [saveMessage, setSaveMessage] = useState(null);
+  const [refresh, setRefresh] = useState(false); //refresh lock(s) meta panel to keep track of locks already added
```
I don't think this is good practice. state should not be control for the component.

julien51 commented on src/assets/js/full-post-page/unlock-content/edit-full-post-page.js:
```diff
@@ -0,0 +1,282 @@
+
+
+  //get current post id
+  const post = useSelect((select) => {
```
If you are getting the post's id here it does not make sense to have the post's locks and ethereumNetworks as props...

julien51 commented on src/assets/js/full-post-page/unlock-content/edit-full-post-page.js:
```diff
@@ -0,0 +1,282 @@
+
+
+  //fetch locks attributes
+  const fetchLockSettings = async (post_id) => {
```
Please use the standard  methods getEditedPostAttribute, editPost

julien51 reviewed (commented):
@SolomonFoskaay I will submit a PR that fixes most of this/ We will pay the bounty and move on. Thanks

julien51 committed:
simplification and cleanup

julien51 committed:
Merge branch 'UPWPlugin08a-Fixed-Issues-4-Locking-Full-Posts-Pages-with-new-UnlockP-flow' of github.com:SolomonFoskaay/unlock-wordpress-plugin into UPWPlugin08a-Fixed-Issues-4-Locking-Full-Posts-Pages-with-new-UnlockP-flow

julien51 committed:
ignoring build files

julien51 committed:
removed files that are not needed

julien51 commented on src/assets/js/admin/utils.js:
```diff
@@ -0,0 +1,49 @@
+ * 
+ * @returns 
+ */
+export const getEthereumNetworksFromSettings = () => {
```
I suspect there exists a native WP function for this. we should use it.

julien51 committed:
removed build files

julien51 committed:
no need to create our own routes

julien51 reviewed (commented):
ok this is WIP. I will fix more stuff later this week.

julien51 committed:
minor fixes

SolomonFoskaay commented:
> @SolomonFoskaay I will submit a PR that fixes most of this/ We will pay the bounty and move on. Thanks

@julien51 Thanks and really appreciate it including the extra fixes you did too.
I will drop my wallet via discord for the bounty payment.

Thanks!

julien51 committed:
renamed for consistency

julien51 committed:
more refactor

julien51 committed:
Merge branch 'master' into UPWPlugin08a-Fixed-Issues-4-Locking-Full-Posts-Pages-with-new-UnlockP-flow

julien51 committed:
removed unlock-wordpress-plugin/languages/unlock-protocol.pot

julien51 committed:
DRYing up code

julien51 committed:
preparing dry up content

julien51 committed:
DRY-ing up code

julien51 commented:
Alright @SolomonFoskaay I think this is the bulk of the rewrite.

Would you be kind enough to do some extensive testing to make sure this works as expected?  I think reviewing this will also be helpful for you to understand a bit more of what was expected?

I also want to rename a few files and I'll do that tomorrow!

After that we're good to go and we'll pay the bounty!

SolomonFoskaay commented:
> Alright @SolomonFoskaay I think this is the bulk of the rewrite.
> 
> Would you be kind enough to do some extensive testing to make sure this works as expected? I think reviewing this will also be helpful for you to understand a bit more of what was expected?
> 
> I also want to rename a few files and I'll do that tomorrow!
> 
> After that we're good to go and we'll pay the bounty!

@julien51 Weldone.

I have tested it as requested. Kindly see below my comments:

Full post/page lock redesigned:
![Selection_090](https://user-images.githubusercontent.com/83863629/232661474-694e8ce4-a315-42e8-bf37-48c2827102fb.png)

Block lock redesigned:
![Selection_092](https://user-images.githubusercontent.com/83863629/232664436-22e2208d-fe7e-4685-ac67-4db3ae3e7306.png)



1. I did observe that you like to reuse code logic with the same feature to avoid duplicating them which is cool as seen in removing the same features codes from edit.js of block lock and edit.js of full post/page lock to have them in one file `admin-locks.js`.

2. You redesigned the lock options for both the block lock and full post/page lock in the post/page editor options panel.

3. You also did some adjustments in the main plugin PHP files

All the above are part of my personal observation that will surely ensure I have a much acceptable code structure to contribute in line with your expectations in other issues fixes that I will be working on for Unlock Protocol Wordpress plugin after this. 

NOTE: Changing the lock options design entirely as you eventually did even for the existing block lock feature I did not work on at all may not have crossed my mind to be acceptable to you before now, considering your initial instruction was to replicate the existing code base style used for the block lock feature.

What we both need to learn here is that it will be best for me to ask for more detailed instructions on how precisely you want a feature to look and work and not just fix them in my own style I deem fit. And, `kindly be ready to support me with detailed instructions on all pending issues that I will be working on next` like issue #45 (though have already worked on it but now have started refactoring the code to align with this code adjustment you did here and will still post some questions in its issue page for clarity before pushing the PR after this PR is merged)

This will reduce the time spent on each PR moving forward.

Really appreciate the code refactoring you did as well.

Thanks!

julien51 commented:
> I did observe that you like to reuse code logic with the same feature to avoid duplicating them which is cool as seen in removing the same features codes from edit.js of block lock and edit.js of full post/page lock to have them in one file admin-locks.js.

Yes, a key characteristic of a maintainable code-base is indeed to re-use code.

> You redesigned the lock options for both the block lock and full post/page lock in the post/page editor options panel.

I did not really redesigned it, just make sure it was consistent by using the same code on both sides because it is the same logic.
